cmd/logger.go

```diff
@@ -137,7 +93,58 @@ func getSource() string {
 	return fmt.Sprintf("[%s:%d:%s()]", filename, lineNum, funcName)
 }
 
-func logIf(level logrus.Level, source string, err error, msg string, data ...interface{}) {
+// TracePoint is to capture stack trace
+type TracePoint struct {
+	Line     int    `json:"line,omitempty"`
+	Filename string `json:"file,omitempty"`
+	Function string `json:"func,omitempty"`
+}
+
+// String method - returns the string
+// representation of a TracePoint instance
+func (tp TracePoint) String() string {
+	return tp.Filename + ":" + strconv.Itoa(tp.Line) + ":" + tp.Function
+}
+
+// trace method - to create stack trace
+func trace(level int) []string {
+	var stack []string
+	var pc uintptr
+	var file string
+	var line int
+	ok := true
+	var callTrace []TracePoint
+	pc, file, line, ok = runtime.Caller(level)
+	rootPath := filepath.Dir(file)
+	for ok {
+		// Clean up the common root path, rootPath,
+		// from the collected stack trace
+		function := runtime.FuncForPC(pc).Name()
+		_, function = filepath.Split(function)
+		file = strings.TrimPrefix(file, rootPath+string(os.PathSeparator))
+		tp := TracePoint{}
+		tp = TracePoint{Line: line, Filename: file, Function: function}
+		callTrace = append(callTrace, tp)
+		level++
+		pc, file, line, ok = runtime.Caller(level)
+	}
+	for _, elem := range callTrace[:len(callTrace)-2] {
+		stack = append(stack, elem.String())
+	}
+	return stack
+}
+
+func reverse(strSlice []string) []string {
+	last := len(strSlice) - 1
+	for i := 0; i < len(strSlice)/2; i++ {
+		strSlice[i], strSlice[last-i] = strSlice[last-i], strSlice[i]
+	}
+	return strSlice
+}
+
+func logIf(level, source, msg string,
```

Comment:
`msg` is a template for `data`.  It should be followed next to each other (like in `Printf()`)

Reply:
The only reason it was written this way was to gather all `string` type arguments together.

I rearranged them so that `msg` appears just before `data`:
`func logIf(level levelType, err error, source, msg string,
	data ...interface{}) {`